server/protocols/webadmin/webadmin-mailrepository/src/main/java/org/apache/james/webadmin/dto/MailDto.java

```diff
@@ -19,39 +19,178 @@
+    public static MailDto fromMail(Mail mail, List<AdditionalFields> additionalFields) throws MessagingException, MissingRequestedField {
         return new MailDto(mail.getName(),
             Optional.ofNullable(mail.getSender()).map(MailAddress::asString),
             mail.getRecipients().stream().map(MailAddress::asString).collect(Guavate.toImmutableList()),
             Optional.ofNullable(mail.getErrorMessage()),
-            Optional.ofNullable(mail.getState()));
+            Optional.ofNullable(mail.getState()),
+            Optional.ofNullable(mail.getRemoteHost()),
+            Optional.ofNullable(mail.getRemoteAddr()),
+            Optional.ofNullable(mail.getLastUpdated()),
+            fetchAttributes(additionalFields, mail),
+            fetchPerRecipientsHeaders(additionalFields, mail),
+            fetchHeaders(additionalFields, mail),
+            fetchBody(additionalFields, mail),
+            fetchMessageSize(additionalFields, mail));
+    }
+
+    private static Optional<Long> fetchMessageSize(List<AdditionalFields> additionalFields, Mail mail) throws MissingRequestedField {
+        if (!additionalFields.contains(AdditionalFields.MESSAGE_SIZE)) {
+            return Optional.empty();
+        }
+        try {
+            return Optional.ofNullable(new Long(mail.getMessageSize()));
+        } catch (MessagingException e) {
+            throw new MissingRequestedField("messageSize");
+        }
+    }
+
+    private static Optional<String> fetchBody(List<AdditionalFields> additionalFields, Mail mail) throws MissingRequestedField {
+        if (!additionalFields.contains(AdditionalFields.BODY)) {
+            return Optional.empty();
+        }
+        try {
+            return Optional.ofNullable(mail.getMessage().getContent().toString());
+        } catch (IOException | MessagingException | NullPointerException e) {
+            throw new MissingRequestedField("body");
+        }
+    }
+
+    private static Optional<Map<String, List<String>>> fetchHeaders(List<AdditionalFields> additionalFields, Mail mail) throws MissingRequestedField {
+        if (!additionalFields.contains(AdditionalFields.HEADERS)) {
+            return Optional.empty();
+        }
+        Map<String, List<String>> headers = new HashMap<>();
+
+        Enumeration<Header> rawHeaders;
+        try {
+            rawHeaders = mail.getMessage().getAllHeaders();
+            while (rawHeaders.hasMoreElements()) {
+                Header header = rawHeaders.nextElement();
+                String name = header.getName();
+                if (!headers.containsKey(name)) {
+                    headers.put(name, new ArrayList<>());
+                }
+
+                headers.get(name).add(header.getValue());
+            }
+
+            return Optional.of(headers);
+        } catch (NullPointerException | MessagingException e) {
+            throw new MissingRequestedField("headers");
```

Comment:
when throwing after catching an exception, it's important to keep the original exception